fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/QuantifiedApply2OuterJoinRule.java

```diff
@@ -29,22 +55,361 @@ public boolean check(OptExpression input, OptimizerContext context) {
+                    Lists.newArrayList(distinctAggregateOutputs), Maps.newHashMap()), distinctConsume);
+        }
+
+        /*
+         * Aggregate(count(1), count(in column), group by correlation columns)
+         *          |
+         *      CTEConsume
+         */
+        private OptExpression buildCountAndConsume() {
+            // CTE count consume
+            Map<ColumnRefOperator, ColumnRefOperator> countConsumeOutputMaps = Maps.newHashMap();
+
+            List<ColumnRefOperator> countAggregateGroupBys = Lists.newArrayList();
+            List<ColumnRefOperator> countAggregateCounts = Lists.newArrayList();
+
+            for (int columnId : inPredicateUsedRefs.getColumnIds()) {
+                ColumnRefOperator ref = factory.getColumnRef(columnId);
+                ColumnRefOperator cr = factory.create(ref, ref.getType(), ref.isNullable());
+                countConsumeOutputMaps.put(cr, ref);
+
+                countAggregateCounts.add(cr);
+            }
+
+            for (int columnId : correlationPredicateInnerRefs.getColumnIds()) {
+                ColumnRefOperator ref = factory.getColumnRef(columnId);
+                ColumnRefOperator cr = factory.create(ref, ref.getType(), ref.isNullable());
+
+                countConsumeOutputMaps.put(cr, ref);
+
+                countAggregateGroupBys.add(cr);
+            }
+
+            OptExpression countConsume =
+                    OptExpression.create(new LogicalCTEConsumeOperator(cteId, countConsumeOutputMaps));
+
+            // rewrite cross/left join on-clause by cte consume output
+            Map<ColumnRefOperator, ScalarOperator> countRewriteMap = Maps.newHashMap();
+            countConsumeOutputMaps.forEach((k, v) -> countRewriteMap.put(v, k));
+            ReplaceColumnRefRewriter rewriter = new ReplaceColumnRefRewriter(countRewriteMap);
+            countJoinOnPredicate = rewriter.rewrite(correlationPredicate);
+
+            // count aggregate
+            Map<ColumnRefOperator, CallOperator> aggregates = Maps.newHashMap();
+            CallOperator countRowsFn = SubqueryUtils.createCountRowsOperator();
+            countRows = factory.create("countRows", countRowsFn.getType(), countRowsFn.isNullable());
+            aggregates.put(countRows, countRowsFn);
+
+            Preconditions.checkState(countAggregateCounts.size() == 1);
+
+            CallOperator countNullFn = SubqueryUtils.createCountRowsOperator(countAggregateCounts.get(0));
+            countNulls = factory.create("countNulls", countNullFn.getType(), countNullFn.isNullable());
+            aggregates.put(countNulls, countNullFn);
+
+            return OptExpression.create(
+                    new LogicalAggregationOperator(AggType.GLOBAL, countAggregateGroupBys, aggregates), countConsume);
+        }
+
+        /*
+         * 1. build cross join if un-correlation sub-query
+         * 2. build left join if correlation sub-query
```

Comment:
Please add a note explaining why there is a difference between cross and left outer